File: src/main/java/com/janrain/backplane/server2/Channel.java
package com.janrain.backplane.server2;

import com.janrain.backplane.common.ExternalizableCore;
import com.janrain.backplane.common.RandomUtils;
import com.janrain.commons.message.MessageException;
import com.janrain.commons.message.MessageField;

import java.util.EnumSet;
import java.util.LinkedHashMap;
import java.util.Map;
import java.util.Set;

/**
 * @author Johnny Bufu
 */
public class Channel extends ExternalizableCore {

    @SuppressWarnings("UnusedDeclaration")
    public Channel() { }

    public Channel(String channelId, BusConfig2 busConfig, int channelExpireSeconds) throws MessageException {
        String id = channelId == null ? RandomUtils.randomString(CHANNEL_NAME_LENGTH) : channelId;
        Map<String,String> data = new LinkedHashMap<String, String>();
        data.put(ChannelField.ID.getFieldName(), id);
        data.put(ChannelField.BUS.getFieldName(), busConfig.getIdValue());
        data.put(ChannelField.EXPIRE_SECONDS.getFieldName(), Integer.toString(channelExpireSeconds));
        data.put(ChannelField.MESSAGE_EXPIRE_DEFAULT_SECONDS.getFieldName(), Integer.toString(busConfig.getRetentionTimeSeconds()));
        data.put(ChannelField.MESSAGE_EXPIRE_MAX_SECONDS.getFieldName(), Integer.toString(busConfig.getRetentionTimeStickySeconds()));
        super.init(id, data);
    }

    @Override
    public String getIdValue() {
        return get(ChannelField.ID);
    }

    @Override
    public Set<? extends MessageField> getFields() {
        return EnumSet.allOf(ChannelField.class);
    }

    public static enum ChannelField implements MessageField {

        // - PUBLIC

        ID("id"),

        BUS("bus"),

        EXPIRE_SECONDS("expire_seconds") {
            @Override
            public void validate(String value) throws MessageException {
                super.validate(value);
                validateInt(getFieldName(), value);
        }},

        MESSAGE_EXPIRE_DEFAULT_SECONDS("message_expire_default_seconds") {
            @Override
            public void validate(String value) throws MessageException {
                super.validate(value);
                validateInt(getFieldName(), value);
        }},

        MESSAGE_EXPIRE_MAX_SECONDS("message_expire_max_seconds") {
            @Override
            public void validate(String value) throws MessageException {
                super.validate(value);
                validateInt(getFieldName(), value);
        }};

        @Override
        public String getFieldName() {
            return fieldName;
        }

        @Override
        public boolean isRequired() {
            return true;
        }

        @Override
        public void validate(String value) throws MessageException {
            if (isRequired()) validateNotBlank(getFieldName(), value);
        }

        // - PRIVATE

        private String fieldName;

        private ChannelField(String fieldName) {
            this.fieldName = fieldName;
        }
    }

    private static final int CHANNEL_NAME_LENGTH = 32;

    private static final long serialVersionUID = 4892206445886794360L;
}

File: src/main/scala/com/janrain/backplane/server2/model/Channel.scala
package com.janrain.backplane.server2.model

import com.janrain.backplane.common.model.{MessageField, MessageFieldEnum, Message}
import scala.collection.JavaConversions._
import com.janrain.util.RandomUtils
import com.janrain.backplane.dao.LegacySupport

/**
 * Binds a previously server-generated channel ID to a bus/config.
 * Also passes along the bus expiration for
 *
 * @author Johnny Bufu
 */
class Channel(data: Map[String,String]) extends Message(data, ChannelFields.values) with LegacySupport[com.janrain.backplane2.server.Channel] {

  def this(channelId: String, config: BusConfig2, channelExpireSeconds: Int) {
    this(Map(
      ChannelFields.ID.name -> (if (channelId == null) RandomUtils.randomString(Channel.CHANNEL_NAME_LENGTH) else channelId),
      ChannelFields.BUS.name -> config.id,
      ChannelFields.EXPIRE_SECONDS.name -> channelExpireSeconds.toString,
      ChannelFields.MESSAGE_EXPIRE_DEFAULT_SECONDS.name -> config.retentionTimeSeconds.toString,
      ChannelFields.MESSAGE_EXPIRE_MAX_SECONDS.name -> config.retentionTimeStickySeconds.toString
    ))
  }

  def idField = ChannelFields.ID

  def asLegacy = new com.janrain.backplane2.server.Channel(mapAsJavaMap(this))
}

object Channel {
  final val CHANNEL_LEGACY_NAME_LENGTH = 32
  final val CHANNEL_NAME_LENGTH = 34
}

object ChannelFields extends MessageFieldEnum {

  type ChannelField = EnumVal

  sealed trait EnumVal extends Value with MessageField {
    def required = true
  }

  val ID = new ChannelField { def name = "id" }

  val BUS = new ChannelField { def name = "bus" }

  val EXPIRE_SECONDS = new ChannelField { def name = "expire_seconds"
    override def validate(fieldValue: Option[String], wholeMessage: Message[_]) {
      super.validate(fieldValue, wholeMessage)
      validateLong(fieldValue)
    }
  }

  val MESSAGE_EXPIRE_DEFAULT_SECONDS = new ChannelField { def name = "message_expire_default_seconds"
    override def validate(fieldValue: Option[String], wholeMessage: Message[_]) {
      super.validate(fieldValue, wholeMessage)
      validateLong(fieldValue)
    }
  }

  val MESSAGE_EXPIRE_MAX_SECONDS = new ChannelField { def name = "message_expire_max_seconds"
    override def validate(fieldValue: Option[String], wholeMessage: Message[_]) {
      super.validate(fieldValue, wholeMessage)
      validateLong(fieldValue)
    }
  }

}
